### PR TITLE
hubble: Add an interface for Parser struct

### PR DESCRIPTION
**Pull request by anubhabMajumdar**
# Description

Currently
- the `Parser` module is missing an interface
- The `LocalObserverServer` struct uses the `Parser` struct directly, instead of using an interface. This is in contrast to the other parameters it accepts like `NamespaceManager`, which is an interface.

# What this change does?

Submitting a minor fix to
- add an interface called `Decoder` for the `Parser` module
- use the `Decoder` interface instead of the `Parser` struct

# Testing done

```bash
$ make build-container
...
<succeeds>
```

Ran unit tests locally.

Fixes: N/A

Signed-off-by: Anubhab Majumdar <anmajumdar@microsoft.com>

### CODE REVIEW

**Review by ghost — APPROVED** [review]
Looks good to me. I edited the PR title a tiny bit to make it clearer.

File: pkg/hubble/observer/local_observer.go
Comment (rolinh) [inline pkg/hubble/observer/local_observer.go:200]: I wouldn't change it here and follow the principle of accepting interfaces and returning structs.

**Review by rolinh — DISMISSED** [review]
Thanks for your contribution! See minor comment below.

File: pkg/hubble/parser/parser.go
Comment (marqc) [inline pkg/hubble/parser/parser.go:30]: nit: missing docs on type and exported method

File: pkg/hubble/observer/local_observer.go
Comment (kaworu) [reply pkg/hubble/observer/local_observer.go:200]: @rolinh IIUC `LocalObserver` doesn't know the actual type though, since after this patch it is constructed with the interface given to `NewLocalServer`.

**Review by kaworu — CHANGES_REQUESTED** [review]
Thanks for the PR @anubhabMajumdar, patch LGTM besides the missing docstring on the `Decoder` interface and it's `Decode` func.

**Comment by squeed** [comment]
poke @anubhabMajumdar  - a few small changes and this seems good-to-go :-)

**Comment by anubhabMajumdar** [comment]
> Thanks for the PR @anubhabMajumdar, patch LGTM besides the missing docstring on the `Decoder` interface and it's `Decode` func.

Hey, sorry for late reply. I am going to address the feedback and update the PR. Thanks for all the feedback @kaworu  @marqc @rolinh

File: pkg/hubble/observer/local_observer.go
Comment (anubhabMajumdar) [reply pkg/hubble/observer/local_observer.go:200]: Yeah, now the LocalObserver constructor accepts the interface to construct the object. It doesn't accept a struct any longer. Similar to NamespaceManager.

Also, the GetPayloadParser is not a member function of the parser class. The New method does return the struct (L41 in parser.go).

File: pkg/hubble/parser/parser.go
Comment (anubhabMajumdar) [reply pkg/hubble/parser/parser.go:30]: Where can I update the docs? I have added a comment on the Decoder interface.

**Comment by anubhabMajumdar** [comment]
@kaworu  @rolinh 
I have updated PR per your feedback; can you please take another look? Thank you!

File: pkg/hubble/parser/parser.go
Comment (marqc) [reply pkg/hubble/parser/parser.go:30]: That's what I meant. All exported names should have comments describing their purpose. Thanks

File: pkg/hubble/parser/parser.go
Comment (marqc) [inline pkg/hubble/parser/parser.go:32]: It may seem unnecessary when interface has only one method, but the exported method also should have a comment describing it's purpose, ie.

```suggestion
	// Decode transforms a monitor event into a hubble event.
	Decode(monitorEvent *observerTypes.MonitorEvent) (*v1.Event, error)
```

**Review by marqc — APPROVED** [review]

**Comment by anubhabMajumdar** [comment]
@marqc Committed your suggested change. I think the CI pipeline to build the image failed with an unrelated error (may be intermittent). I tried building the container images locally, and it works as expected.

**Comment by marqc** [comment]
@anubhabMajumdar I'm not a maintainer, but you have my LGTM  ;)

I can see that there are still pending change requests on this PR. If you addressed them, please re-request reviews from kaworu and rolinh.

**Comment by anubhabMajumdar** [comment]
> @anubhabMajumdar I'm not a maintainer, but you have my LGTM ;)
> 
> I can see that there are still pending change requests on this PR. If you addressed them, please re-request reviews from kaworu and rolinh.

@kaworu  @rolinh Can you help me check this in? Thank you!

**Review by kaworu — APPROVED** [review]
Thanks @anubhabMajumdar!

**Comment by joestringer** [comment]
Hi @anubhabMajumdar , I see that you have rebased+updated the PR. In cases where there is an unresolved conflict with the `main` branch or the PR is significantly out-of-date (perhaps a week or more), this can be useful to sync back up to help drive the PR forward.

At the same time, I did want to ensure that you're aware that Cilium is a very fast moving project, and GitHub will report that your PRs are "out-of-date with the base branch" almost all the time. Generally speaking this is OK, PRs which are out-of-date with the base branch can still be nominated to merge as long as the CI is passing and required codeowners have approved the PR. After we click the merge button, the PR goes into the merge queue and there will be some sanity checks run against the PR to minimize the likelihood of breakage due to the temporary divergence from upstream `main`. Just be aware that if you do update/refresh the PR, then this will clear the CI results so we'll need to retrigger the tests afresh each time. We're happy to do this for you until you [become an organization member](https://github.com/cilium/community/blob/main/CONTRIBUTOR-LADDER.md#organization-member), feel free to reach out to us on #development channel on Slack if you need tests to be retriggered or you'd like help trying to understand test results.

**Comment by anubhabMajumdar** [comment]
> Hi @anubhabMajumdar , I see that you have rebased+updated the PR. In cases where there is an unresolved conflict with the `main` branch or the PR is significantly out-of-date (perhaps a week or more), this can be useful to sync back up to help drive the PR forward.
> 
> At the same time, I did want to ensure that you're aware that Cilium is a very fast moving project, and GitHub will report that your PRs are "out-of-date with the base branch" almost all the time. Generally speaking this is OK, PRs which are out-of-date with the base branch can still be nominated to merge as long as the CI is passing and required codeowners have approved the PR. After we click the merge button, the PR goes into the merge queue and there will be some sanity checks run against the PR to minimize the likelihood of breakage due to the temporary divergence from upstream `main`. Just be aware that if you do update/refresh the PR, then this will clear the CI results so we'll need to retrigger the tests afresh each time. We're happy to do this for you until you [become an organization member](https://github.com/cilium/community/blob/main/CONTRIBUTOR-LADDER.md#organization-member), feel free to reach out to us on #development channel on Slack if you need tests to be retriggered or you'd like help trying to understand test results.

Thanks for the explanation, Joe. I was thinking it may be necessary to keep rebasing my PR so that it can be merged by maintainers. Will reach out on Slack if I need help retriggering the tests in future. Cheers!

**Review by rolinh — APPROVED** [review]

**Pull request merged**
